src/main/java/org/jitsi/videobridge/cc/BitrateController.java

```diff
@@ -927,27 +993,42 @@ private synchronized void update()
             logger.debug("Prioritizing endpoints, adjusted last-n: " + adjustedLastN +
                 ", sorted endpoint list: " +
                 conferenceEndpoints.stream().map(AbstractEndpoint::getID).collect(Collectors.joining(", ")) +
-                ". Selected endpoints: " + String.join(", ", selectedEndpointIds) +
-                ". Pinned endpoints: " + String.join(", ", pinnedEndpointIds));
+                ". Endpoints constraints: " + Arrays.toString(videoConstraintsMap.values().toArray()));
         }
 
-        int endpointPriority = 0;
-
-        // First, bubble-up the selected endpoints (whoever's on-stage needs to
-        // be visible).
-        for (Iterator<AbstractEndpoint> it = conferenceEndpoints.iterator();
-             it.hasNext() && endpointPriority < adjustedLastN;)
+        // Before endpoint constraints we had had the notion of selected and
+        // pinned endpoints. "selected" endpoints were endpoints that had
+        // 720p ideal height and prioritized over everything else and were
+        // allocated bandwidth first (up to their preferred resolution).
+        // "Pinned" endpoints where those thee had 180p ideal height and were
+        // ranked lower than selected endpoints but before the rest of the pack.
+        // Finally, any remaining endpoints were ranked after pinned endpoints.
+
+        Map<String, VideoConstraints> videoConstraintsMapCopy = videoConstraintsMap;
+
+        List<EndpointMultiRank> endpointMultiRankList = conferenceEndpoints
+            .stream()
+            .map(endpoint -> {
+                VideoConstraints videoConstraints = videoConstraintsMapCopy
+                    .getOrDefault(endpoint.getID(), defaultVideoConstraints);
+
+                int rank = conferenceEndpoints.indexOf(endpoint);
+                return new EndpointMultiRank(rank, videoConstraints, endpoint);
+            })
+            .sorted(new EndpointMultiRanker())
+            .collect(Collectors.toList());
+
+        for (EndpointMultiRank endpointMultiRank : endpointMultiRankList)
```

Comment:
You don't use the rank other than to access the endpoint. I think you can you just `.map(r -> r.endpoint)` after sorting the stream above and then work with the endpoints.